scripts/build_utils.sh

```diff
@@ -1694,54 +1690,157 @@ pr_only_for() {
+}
+
+match_filename() {
+  local f=$1
+  local p=$2
+  if [[ $p =~ .*/.+ ]]; then
+    # patterns with / in front or middle are absolute matches
+    if match_file_or_directory "$f" "$p"; then return 0; fi
+  else
+    # otherwise we can match under subdirectories too
+    if match_file_or_directory "$f" "$p"; then return 0; fi
+    if match_file_or_directory "$f" ".+/$p"; then return 0; fi
+  fi
+  return 1
+}
+
+no_filenames_match() {
+  local files=$1
+  local patterns=$(codeowners_patterns_to_regex "$2")
+  for f in $files; do
+    for p in $patterns; do
+      # add leading slash to simplify matching
+      if match_filename "/$f" "$p"; then
+        echo "filename '$f' matched pattern '$p'"
+        return 1
+      fi
+    done
+  done
+  return 0
+}
+
+all_filenames_match() {
+  local files=$1
+  local patterns=$(codeowners_patterns_to_regex "$2")
   for f in $files; do
+    # 0 is true, 1 is false
     local match=1
-    for p in "${local_patterns[@]}"; do
+    for p in $patterns; do
       # pattern loop: if one pattern matches, skip the others
-      if [[ $f == $p ]]; then match=0; break; fi
+      # add leading slash to simplify matching
+      if match_filename "/$f" "$p"; then match=0; break; fi
     done
     # file loop: if this file matched no patterns, the group fails
     # (one mismatch spoils the whole bushel)
-    if [[ $match -eq 1 ]] ; then all_match=1; break; fi
+    if [[ $match -eq 1 ]] ; then
+      echo "filename '$f' matched no patterns"
+      return 1
+    fi
   done
-  return $all_match
+  return 0
 }
 
 docs_pr_only() {
   DOCS_ONLY=false
-  local patterns=(
-    'doc/*'
-    'admin/*'
-    'src/sample.ceph.conf'
-    'CodingStyle'
-    '*.rst'
-    '*.md'
-    'COPYING*'
-    'README.*'
-    'SubmittingPatches'
-    '.readthedocs.yml'
-    'PendingReleaseNotes'
-  )
-  if pr_only_for patterns; then DOCS_ONLY=true; fi
+  local filenames=$(pr_filenames_changed)
+  echo -e "changed files:\n$filenames"
+  local patterns='doc/
+admin/
+src/sample.ceph.conf
+CodingStyle
+*.rst
+*.md
+COPYING*
+README.*
+SubmittingPatches
+.readthedocs.yml
+PendingReleaseNotes'
+  if all_filenames_match "$filenames" "$patterns"; then DOCS_ONLY=true; fi
 }
 
 container_pr_only() {
   CONTAINER_ONLY=false
-  local patterns=(
-    'container/*'
-    'Dockerfile.build'
-    'src/script/buildcontainer-setup.sh'
-    'src/script/build-with-container.py'
-  )
-  if pr_only_for patterns; then CONTAINER_ONLY=true; fi
+  local filenames=$(pr_filenames_changed)
+  local patterns='container/
+Dockerfile.build
+src/script/buildcontainer-setup.sh
+src/script/build-with-container.py'
+  if all_filenames_match "$filenames" "$patterns"; then CONTAINER_ONLY=true; fi
+}
+
+# given a CODEOWNERS file whose lines look like this:
+#   /doc/dev/rbd*    @ceph/rbd @ceph/doc-writers
+#
+# return the pattern string of all lines matching the given codeowner
+patterns_for_codeowner() {
+  grep -w "$1" "$WORKSPACE/ceph/.github/CODEOWNERS" | cut -f 1 -d " "
```

Comment:
from https://jenkins.ceph.com/view/all/job/preserve-ceph-pull-requests/1/console against https://github.com/ceph/ceph/pull/63076:
```
++ patterns_for_codeowner @ceph/rbd
++ grep -w @ceph/rbd /home/jenkins-build/build/workspace/preserve-ceph-pull-requests/ceph/.github/CODEOWNERS
++ cut -f 1 -d ' '
grep: /home/jenkins-build/build/workspace/preserve-ceph-pull-requests/ceph/.github/CODEOWNERS: No such file or directory
```
`$WORKSPACE/ceph` doesn't appear to be the right location for the ceph checkout. i had copied this from `pr_only_for()`:
```sh
  if ! [[ "$(git config --get remote.origin.url)" =~ "ceph/ceph.git" ]]; then
    cd "$WORKSPACE/ceph"
  fi
```
but that "cd" doesn't happen in the logs